modules/social_features/social_group/modules/social_group_welcome_message/social_group_welcome_message.module

```diff
@@ -0,0 +1,150 @@
+  ];
+
+  \Drupal::moduleHandler()->alter('social_group_types', $social_group_types);
+
+  $group_forms = [];
+  foreach ($social_group_types as $social_group_type) {
+    $group_forms[] = "group_{$social_group_type}_add_form";
+    $group_forms[] = "group_{$social_group_type}_edit_form";
+  }
+
+  if (
+    in_array($form_id, $group_forms) &&
+    isset($form['private_message_body']) &&
+    isset($form['private_message_send'])
+  ) {
+    $form['private_message_body']['#states'] = [
+      'visible' => [
+        ':input[name="private_message_send[value]"]' => ['checked' => TRUE],
+      ],
+    ];
+  }
+}
+
+/**
+ * Implements hook_ENTITY_TYPE_insert().
+ */
+function social_group_welcome_message_group_content_insert(GroupContentInterface $group_content) {
+  $type = $group_content->getContentPlugin()->getPluginId();
+  $group = $group_content->getGroup();
+
+  // If new membership is created.
+  if (
+    $type === 'group_membership' &&
+    !$group->get('private_message_send')->isEmpty() &&
+    $group->get('private_message_send')->value == 1 &&
+    !$group->get('private_message_body')->isEmpty()
+  ) {
+    $g_type_id = $group->getGroupType()->id();
+
+    // Get group admins entities.
+    $g_admins = $group->getMembers($g_type_id . '-group_admin');
+    $g_admins_users = [];
+    /** @var \Drupal\group\GroupMembership $member */
+    foreach ($g_admins as $key => $member) {
+      $g_admins_users[$key] = $member->getUser();
+    }
+
+    // Get group managers entities.
+    $g_managers = $group->getMembers($g_type_id . '-group_manager');
+    $g_managers_users = [];
+    /** @var \Drupal\group\GroupMembership $member */
+    foreach ($g_managers as $key => $member) {
+      $g_managers_users[$key] = $member->getUser();
+    }
+
+    // Set sender (owner) of the pm.
+    if (count($g_admins_users) === 1) {
+      list($owner) = $g_admins_users;
+    }
+    elseif (count($g_managers_users) > 0) {
+      list($owner) = $g_managers_users;
+    }
+    else {
+      $owner = $group->getOwner();
+    }
+
+    // Merge all managers of the group to managers array.
+    $managers = array_merge($g_admins_users, $g_managers_users, [$group->getOwner()]);
+
+    // Load joined user.
+    $user = $group_content->getEntity();
+
+    // Add the joined user to the already existing managers array.
+    $managers[] = $user;
+
+    /** @var \Drupal\private_message\Service\PrivateMessageServiceInterface $private_message_service */
+    $private_message_service = \Drupal::service('private_message.service');
+
+    // Create a pm thread between these users.
+    $thread = $private_message_service->getThreadForMembers($managers);
+
+    // Get body of pm.
+    $private_message_body = $group->get('private_message_body')->first()->getValue();
+
+    // Create a single message with the pm body.
+    $private_message = \Drupal::entityTypeManager()->getStorage('private_message')->create([
+      'owner' => $owner,
+      'message' => $private_message_body,
+    ]);
```

Comment:
Since we're not using the text format here, perhaps we should not create a field with a text format but use the `string_long` field type instead (this will remove the WYSIWYG though, but we'll now have bugs here either way).

Reply:
Very good point. If I'm not mistaken private messages support some text format (maybe `full html`, maybe just `basic html`). So we really should check this and adjust if needed (I mean if private messages support `full html` text then we can leave this as it is, agree?)

Reply:
I think we should be careful here. By default on sites I think only site managers (and maybe content managers) have access to `full html`. However, normal users can also make groups. So if this is a group setting, then this should probably be adjusted also to what the user creating the welcome message has access to.

However, I suspect that private messages only allows `basic html` or `plain text` (since any user can create a private message), so we should use that, because that'll also be used for rendering (i.e. allowing SM to use `full text` would not render correctly).